lib/src/r_tree/r_tree.dart

```diff
@@ -61,10 +63,285 @@ class RTree<E> {
+
+    // recursively build the tree with the given data from scratch using OMT algorithm
+    Node<E> node = _build(items, 0, items.length - 1, 0);
+
+    if (_root.children.length == 0) {
+      // save as is if tree is empty
+      _root = node;
+    } else if (_root.height == node.height) {
+      // split root if trees have the same height
+      _growTree(_root, node);
+    } else {
+      if (_root.height < node.height) {
+        // swap trees if inserted one is bigger
+        final tmpNode = _root;
+        _root = node;
+        node = tmpNode;
+      }
+
+      // insert the small tree into the large tree at appropriate level
+      _insertTree(_root.height - node.height - 1, node);
+    }
+
+    return;
+  }
+
+  void _insertTree(int level, Node<E> inode) {
+    final List<Node<E>> insertPath = [];
+
+    // find the best node for accommodating the item, saving all nodes along the path too
+    final node = _chooseSubtree(inode, _root, level, insertPath);
+
+    node.children.add(inode);
+    node.updateBoundingRect();
+
+    // split on node overflow; propagate upwards if necessary
+    while (level >= 0) {
+      if (insertPath[level].children.length > _branchFactor) {
+        _split(insertPath, level);
+        level--;
+      } else {
+        break;
+      }
+    }
+
+    // fix all the bounding rectangles along the insertion path
+    insertPath.forEach((e) => e.updateBoundingRect());
+  }
+
+  Node<E> _chooseSubtree(
+      Node<E> inode, Node<E> node, int level, List<Node<E>> path) {
+    while (true) {
+      path.add(node);
+
+      if (node is LeafNode || path.length - 1 == level) {
+        break;
+      }
+
+      final nonLeafNode = node as NonLeafNode<E>;
+
+      num minArea = double.infinity;
+      num minEnlargement = double.infinity;
+      Node<E>? targetNode;
+
+      // no leaves here
+      for (final child in nonLeafNode.children) {
+        if (child is NonLeafNode<E>) {
+          final area = child.rect.width * child.rect.height;
+          final enlargement = inode.expansionCost(child);
+
+          // choose entry with the least area enlargement
+          if (enlargement < minEnlargement) {
+            minEnlargement = enlargement;
+            minArea = area < minArea ? area : minArea;
+            targetNode = child;
+          } else if (enlargement == minEnlargement) {
+            // otherwise choose one with the smallest area
+            if (area < minArea) {
+              minArea = area;
+              targetNode = child;
+            }
+          }
+        }
+      }
+
+      node = targetNode ?? nonLeafNode.children.first;
+    }
+
+    return node;
+  }
+
+  Node<E> _build(List<RTreeDatum<E>> items, int left, int right, int height) {
+    final N = right - left + 1;
+    var M = _branchFactor;
+    Node<E> node;
+
+    if (N <= M) {
+      // reached leaf level; return leaf
+      node =
+          LeafNode(_branchFactor, initialItems: items.sublist(left, right + 1));
+      return node;
```

Comment:
Do we need to assign on the line above since we're returning?

Reply:
nope, good point